### DISCUSSION TITLE
Driving high impedance to signal and Logging Issues

### DISCUSSION THREAD

**Discussion by PiyushSaini09** [cocotb/cocotb#2885] [Q&A]
Hi -

I've got some questions regarding driving high impedance to a signal and logging :-

1) I'm trying to drive a signal (for example - test_signal) with a value other than 1 or 0.
    I tried using `dut.test_signal.value = 'z'`, but throws the following error:
    `Unsupported type for value assignment: <class 'str'> ('z')`. 

    I went through `cocotb.types.Logic` (https://docs.cocotb.org/en/v1.6.0/library_reference.html?highlight=Logic#:~:text=class-,cocotb.types.Logic,-(value%3A%20Optional%5BUnion) but couldn't understand how it can be used to get a high impedance state.
    Is it possible to drive a 'z' to a signal in cocotb or am I doing something wrong?
    
2) In Makefile I've defined environment variable `COCOTB_LOG_LEVEL = ERROR` and I am trying the following snippet:
     `dut._log.info("This is an INFO message")`
      `dut._log.error("This is an ERROR message")`
      
      What I expected was that I will only get the logging messages of `ERROR`. However, I'm getting both `INFO` and `ERROR` 
      messages. Is this a correct behaviour or am I doing something wrong?
      I also tried adding `COCOTB_LOG_LEVEL = DEBUG` and tried a logging message with `debug` but that was never shown unlike 
      the other two. How does debug logging work?

3) I was trying to get a colored logging output for different logging levels by setting the environment variable 
    `COCOTB_ANSI_OUTPUT = 1`. But I did not get the colors on logs. I then tried checking whether the color was requested or not 
    by using `print(cocotb.utils.want_color_output())` to which I got the output as `False`. Does this mean that my simulator is not capable of giving colored logging outputs?

Thank you.

============================================================
Cocotb Version - v1.6.0
Simulator          - Xcelium
Python              - 3.9.0 64-bit

**Comment by ktbarrett** [comment]
1. There is a couple things here. First you cannot "drive" signals from cocotb, if you are trying to write a tri-state state driver, what you are doing won't work. But to deposit a `Z` onto a signal you can use either BinaryValue or Logic `dut.handle.value = BinaryValue("Z")` or `dut.handle.value = Logic("Z")`.

2. This is expected behavior. Messages at *or above* the log level are allowed to appear. Read more in [the python documentation](https://docs.python.org/3/library/logging.html).
 
3. This does likely mean that your simulator doesn't support colored output. @cmarqu knows more about Xcelium and might be able to shine more light.

**Reply by cmarqu** [reply]
I didn't check for some time, but in the past, Xcelium was able to output colors with `COCOTB_ANSI_OUTPUT=1`. (The automatic `isatty()` detection does not work because, in the Makefile, we use `| tee` to capture stdout for the logs.)
Are you sure you are setting the environment variable correctly?

**Reply by PiyushSaini09** [reply]
yes, I'm pretty sure that I'm setting the environment variable correctly.
Here is the snippet of my Makefile
![image](https://user-images.githubusercontent.com/96406295/156298620-10b99945-fce4-42b4-9601-817e8001c590.png)

**Reply by PiyushSaini09** [reply]
Thank you for the help.
I tried both BinaryValue and Logic. BinaryValue **fulfilled** my need but Logic still doesn't seem to work. It threw the following error:
`TypeError: Unsupported type for value assignment: <class 'cocotb.types.logic.Logic'> (Logic('Z'))`

**Reply by ktbarrett** [reply]
COCOTB_LOG_LEVEL and COCOTB_ANSI_OUTPUT are environment variables, so they must be exported. Try the following.
 ```make
export COCOTB_LOG_LEVEL = DEBUG
export COCOTB_ANSI_OUTPUT = 1
```

> `TypeError: Unsupported type for value assignment: <class 'cocotb.types.logic.Logic'> (Logic('Z'))`

Ah, I added support for LogicArrays, but not individual Logics. Sorry for the confusion. That support should be added in 1.7, but it seems like BinaryValue works for now.

**Reply by PiyushSaini09** [reply]
Everything worked as expected. I'm getting the colored logs. 
Thank you for clarification :)

I have a suggestion, can you please update the documentation such that the users knows that they need to use `export` for environment variables?

**Reply by ktbarrett** [reply]
That's a good suggestion, will do.
